Handle the actions when the user taps on an Inbox Note

pmusolino opened the pull request:
Closes: #5955 

## Description
In this PR I implemented the actions when the user taps on a note: we should present a Safari View Controller or the Authenticated web View (only for sites hosted on WP.com) based on the site currently selected. Plus, when the user tap "dismiss" on a note, we need to remove it.
Note that for the dismissing action, I updated also the yosemite layer because it doesn't make sense for us to store notes that we don't need to show anymore and that will be hidden.

**Two important things to consider:**
- While testing, I discovered that the notes we display in the app are not similar to the one displayed on the web. We display more elements because on the web there are some extra filters, so we should align the request. https://github.com/woocommerce/woocommerce-ios/issues/6234
- We should test if the pagination works as expected also after dismissing a note, after that we apply the fix above.

## Testing instructions

#### Case 1
1. Select a self-hosted website.
2. Open the inbox notes.
3. Press the action button (the one in "purple/pink") on a note that in the web is not "actioned" (unread).
4. Go on the web, you should see the note as "read" (actioned).
5. Also, you should see the Safari View Controller opened with the URL of the inbox note.

#### Case 2
1. Select a store hosted on WPCom.
2. Open the inbox notes.
3. Press the action button (the one in "purple/pink") on a note that in the web is not "actioned" (unread).
4. Go on the web, you should see the note as "read" (actioned).
5. Also, you should see the Authenticated WebView opened with the URL of the inbox note.

#### Case 3
1. Open the inbox notes.
2. Press dismiss on a note.
3. You should see the note dismissed on the app and also on the web.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

pmusolino committed:
feat: implemented action when the user taps on a note

pmusolino committed:
fix: removed comment

pmusolino committed:
feat: implemented action `markInboxNoteAsActioned`

pmusolino committed:
feat: implemented action `dismissInboxNote`

pmusolino committed:
feat: removed print

pmusolino committed:
fix: updated `dismissInboxNote` path

pmusolino committed:
fix: indentation

pmusolino committed:
fix: updated `InboxNotesRemoteTests` failing tests

pmusolino committed:
fix: updated `InboxNotesStoreTests` failing tests

pmusolino committed:
feat: improvements to `dismissInboxNote` action adding a completion closure

pmusolino committed:
feat: activity indicator when the dismiss button is pressed.

pmusolino committed:
feat: now the action for dismissing a note deletes the note locally + updated tests

pmusolino committed:
fix: comment

jaclync commented on WooCommerce/Classes/ViewRelated/Inbox/InboxNoteRow.swift:
```diff
@@ -53,29 +57,74 @@ struct InboxNoteRow: View {
+        .wooNavigationBarStyle()
+        }
+        else {
+            SafariSheetView(url: displayedURL ?? WooConstants.URLs.blog.asURL())
```
nit: does it make sense to `guard let` the URL and show an empty view or make a non-optional URL at the beginning of the function to be reused in both cases?
```suggestion
        let isWPComStore = ServiceLocator.stores.sessionManager.defaultSite?.isWordPressStore ?? false
        let url = displayedURL?.absoluteURL ?? WooConstants.URLs.blog.asURL()

        if isWPComStore {
        NavigationView {
            AuthenticatedWebView(isPresented: $showWebView,
                                 url: url,
                                 urlToTriggerExit: nil) {

            }
             .navigationTitle(Localization.inboxWebViewTitle)
             .navigationBarTitleDisplayMode(.inline)
             .toolbar {
                 ToolbarItem(placement: .confirmationAction) {
                     Button(action: {
                         showWebView = false
                     }, label: {
                         Text(Localization.doneButtonWebview)
                     })
                 }
             }
        }
        .wooNavigationBarStyle()
        }
        else {
            SafariSheetView(url: url)
```

jaclync commented on WooCommerce/Classes/ViewRelated/Inbox/InboxNoteRow.swift:
```diff
@@ -7,6 +7,9 @@ struct InboxNoteRow: View {
 
     // Tracks the scale of the view due to accessibility changes.
     @ScaledMetric private var scale: CGFloat = 1
+    @State private var displayedURL: URL?
```
nit: any reasons not just `url` or `actionURL`?

jaclync commented on WooCommerce/Classes/ViewRelated/Inbox/InboxNoteRow.swift:
```diff
@@ -7,6 +7,9 @@ struct InboxNoteRow: View {
     // Tracks the scale of the view due to accessibility changes.
     @ScaledMetric private var scale: CGFloat = 1
+    @State private var displayedURL: URL?
+    @State private var showWebView: Bool = false
```
nit: `show*` sounds like a function, how about `shows*`? or is this a SwiftUI pattern?
```suggestion
    @State private var showsWebView: Bool = false
```

jaclync commented on WooCommerce/Classes/ViewRelated/Inbox/InboxNoteRow.swift:
```diff
@@ -7,6 +7,9 @@ struct InboxNoteRow: View {
     @ScaledMetric private var scale: CGFloat = 1
+    @State private var displayedURL: URL?
+    @State private var showWebView: Bool = false
+    @State private var dismissButtonIsLoading: Bool = false
```
super nit: `is*` might be more clear, but it's really minor and up to you 😄 
```suggestion
    @State private var isDismissButtonLoading: Bool = false
```

jaclync commented on WooCommerce/Classes/ViewRelated/Inbox/InboxNoteRow.swift:
```diff
@@ -53,29 +57,74 @@ struct InboxNoteRow: View {
+                            Button(Localization.dismiss) {
+                                dismissButtonIsLoading = true
+                                viewModel.dismissInboxNote { _ in
+                                    dismissButtonIsLoading = false
```
nit: maybe `dismissButtonIsLoading` state could be provided by the view model? then it can tested to make sure the state is set/reset

jaclync commented on WooCommerce/Classes/ViewRelated/Inbox/InboxNoteRowViewModel.swift:
```diff
@@ -20,7 +21,21 @@ struct InboxNoteRowViewModel: Identifiable, Equatable {
+         locale: Locale = .current,
+         calendar: Calendar = .current,
+         stores: StoresManager = ServiceLocator.stores,
+         storageManager: StorageManagerType = ServiceLocator.storageManager) {
```
nit: is this `storageManager: StorageManagerType` used in the view model?

jaclync commented on WooCommerce/Classes/ViewRelated/Inbox/InboxNoteRowViewModel.swift:
```diff
@@ -34,21 +49,72 @@ struct InboxNoteRowViewModel: Identifiable, Equatable {
+    func markInboxNoteAsActioned() {
+        guard let actionID = actions.first?.id else {
+            return
+        }
```
I've seen notes with more than one action during testing, it'd be safer passing the action ID here:
```suggestion
    func markInboxNoteAsActioned(actionID: Int64) {
```

jaclync commented on WooCommerce/Classes/ViewRelated/Inbox/InboxNoteRowViewModel.swift:
```diff
@@ -34,21 +49,72 @@ struct InboxNoteRowViewModel: Identifiable, Equatable {
+                                                              noteID: id,
+                                                              actionID: actionID) { result in
+            switch result {
+            case .success(_):
```
super nit:
```suggestion
            case .success:
```

jaclync commented on Yosemite/Yosemite/Stores/InboxNotesStore.swift:
```diff
@@ -73,20 +73,20 @@ private extension InboxNotesStore {
-                self?.upsertStoredInboxNotesInBackground(readOnlyInboxNotes: [inboxNote], siteID: siteID, shouldDeleteExistingNotes: false) {
-                    completion(.success(inboxNote))
-                }
+            case .success(_):
```
super nit:
```suggestion
            case .success:
```

jaclync commented on Yosemite/YosemiteTests/Stores/InboxNotesStoreTests.swift:
```diff
@@ -157,13 +157,13 @@ final class InboxNotesStoreTests: XCTestCase {
         // Given a stubbed inbox note network response
         let sampleInboxNoteID: Int64 = 296
-        network.simulateResponse(requestUrlSuffix: "admin/notes/\(sampleInboxNoteID)", filename: "inbox-note")
+        network.simulateResponse(requestUrlSuffix: "admin/notes/delete/\(sampleInboxNoteID)", filename: "inbox-note")
```
nit: maybe an assert on storage inbox notes would be useful for comparing this and the next test case:
```suggestion
        network.simulateResponse(requestUrlSuffix: "admin/notes/delete/\(sampleInboxNoteID)", filename: "inbox-note")
        XCTAssertEqual(storedInboxNotesCount, 0)
```

jaclync commented on Yosemite/YosemiteTests/Stores/InboxNotesStoreTests.swift:
```diff
@@ -174,19 +174,44 @@ final class InboxNotesStoreTests: XCTestCase {
+            self.store.onAction(action)
+        }
+
+        // Then no inbox notes should be stored
```
I like the more detailed `Given/When/Then` comments 👍

jaclync reviewed (commented):
Thanks for noticing and creating an issue for the extra filters for the notes request in core!

Code LGTM, but I ran into an issue where the webview from every action shows the WooCommerce blog (the default URL when the `displayedURL` is `nil`). After some debugging, I saw [this SO post](https://stackoverflow.com/a/64140190/9185596) and changed the sheet action in [this branch](https://github.com/woocommerce/woocommerce-ios/compare/issue/5955-handle-the-action-when-the-user-taps-on-an-inbox-note...jaclync/pr-testing?expand=1) and it seems to be working now. This is quite unexpected, I'm not sure why a `@State` variable inside `@ViewBuilder` call isn't up-to-date. Have you seen this before, and do you know of other solutions?

pmusolino committed:
Update WooCommerce/Classes/ViewRelated/Inbox/InboxNoteRow.swift

Co-authored-by: Jaclyn Chen <watertranquil@gmail.com>

pmusolino committed:
refactor: variable name `actionURL`

pmusolino committed:
refactor: showsWebView variable name

pmusolino committed:
refactor: variable name `isDismissButtonLoading`

pmusolino committed:
fix: removed unused `storageManager`

pmusolino committed:
pass: action id in `markInboxNoteAsActioned` method

pmusolino committed:
fix: code style

pmusolino committed:
feat: an assert on storage inbox notes is useful for comparing this and the next test case

pmusolino committed:
fix: changed the sheet action

pmusolino commented:
> Code LGTM, but I ran into an issue where the webview from every action shows the WooCommerce blog (the default URL when the displayedURL is nil). After some debugging, I saw [this SO post](https://stackoverflow.com/a/64140190/9185596) and changed the sheet action in [this branch](https://github.com/woocommerce/woocommerce-ios/compare/issue/5955-handle-the-action-when-the-user-taps-on-an-inbox-note...jaclync/pr-testing?expand=1) and it seems to be working now. This is quite unexpected, I'm not sure why a @State variable inside @ViewBuilder call isn't up-to-date. Have you seen this before, and do you know of other solutions?

Wow, Good catch. I never experienced a similar case, but I updated the code following your suggestions here: 83b1f6dc9d18d99cd24dba2b73ea5462aa5d4abf

About the other suggestions, I updated my code in different commits 👍

pmusolino commented:
@jaclync ready for a second check 👀

jaclync commented on WooCommerce/Classes/ViewRelated/Inbox/InboxNoteRowViewModel.swift:
```diff
@@ -1,5 +1,6 @@
 import SwiftUI
 import Yosemite
+import protocol Storage.StorageManagerType
```
nit: I think this import isn't needed anymore 😄

jaclync reviewed (approved):
Thanks for the updates @pmusolino! LGTM :shipit: 

Some minor issues I noticed with a **WP.com test site**:
- In iOS 15.2 iPhone 11 simulator, opening a webview always requires a verification code for 2FA every time (not an issue on iOS 14.5 device)
- Tapping "Done" in the webview does nothing (reproducible in device and simulator)
<img src="https://user-images.githubusercontent.com/1945542/155077635-307cfd7d-a6e5-4454-9ab4-9deca28ba015.png" width="300" />

pmusolino committed:
fix: removed unused import

pmusolino committed:
fix: authenticated web view dismiss

pmusolino commented:
> In iOS 15.2 iPhone 11 simulator, opening a webview always requires a verification code for 2FA every time (not an issue on iOS 14.5 device)

You should disable AutoProxxy, is a known issue (sorry for not mentioning it!). More info here: pMz3w-bAI#comment-80199

> Tapping "Done" in the webview does nothing (reproducible in device and simulator)

Fixed here 416367897aa74c71c4791c289398e368d7ddb529
